### PR TITLE
docs: add explanation for providedIn any

### LINKED ISSUE #35179: document providedIn: any and platform
# 📚 Docs or angular.io bug report

### Description

Angular 9 introduces providedIn: any and platform for provided scope, which is not documented.

### What's the affected URL?**
https://next.angular.io/guide/providers#provider-scope

## 🔥 Exception or Error
<pre><code>

</code></pre>

### PR DESCRIPTION
**Pull request by santoshyadavdev**
Fixes #35179

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #35179

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by santoshyadav198613** [comment]
Hi @gkalpak and @kapunahelewong ,

Do we need to do something different for images, the error is due to path of image.

File: aio/content/guide/providers.md [hunk 1]
```diff
@@ -52,6 +52,14 @@ Any component created within a lazy loaded module’s context, such as by router
 
 Though you can provide services by lazy loading modules, not all services can be lazy loaded. For instance, some modules only work in the root module, such as the Router. The Router works with the global location object in the browser.
 
+With Angular 9, it is possible to provide a new instance of service with each lazy loaded module. You can change the UserService code like:
```
Comment (kapunahelewong) [inline aio/content/guide/providers.md]:
```suggestion
With Angular version 9, you can provide a new instance of service with each lazy loaded module. The change to `UserService` would be as follows:
```

File: aio/content/guide/providers.md [hunk 2]
```diff
@@ -52,6 +52,14 @@ Any component created within a lazy loaded module’s context, such as by router
...
+
+<code-example path="providers/src/app/user.service.2.ts"  header="src/app/user.service.ts"></code-example>
+
+Remember this is available for User defined services, and all eager loaded module will share an singleton insance, as shown in image below.
```
Comment (kapunahelewong) [inline aio/content/guide/providers.md]:
```suggestion
Using `providedIn: 'any'` is available for user defined services, or services.... . 
With this configuration, all eagerly loaded modules share a singleton instance, however lazy loaded modules each get their own unique instance, as in the following diagram.
```
Could you define user defined services?

**Review by kapunahelewong — CHANGES_REQUESTED** [review]
Thank you so much, @santoshyadav198613! Just a couple of suggestions on the writing. Regarding the image, could you use colors and styles like the ones in these images? 
https://angular.io/generated/images/guide/dependency-injection/injectors.svg
https://angular.io/generated/images/guide/inputs-outputs/input.svg
https://angular.io/generated/images/guide/inputs-outputs/input-diagram-target-source.svg
https://angular.io/generated/images/guide/inputs-outputs/input-output-diagram.svg

Those are all relatively recent additions where we're trying to use the Angular colors. If your software doesn't have what you need to do it, I'm happy to help you with it. Regarding the content of the diagram, could you also explain in the doc what's in each (like what's the `ConfigService`, etc.). If we have diagrams, we need to have a full explanations so that the content is still there for those using assistive technology like screen readers or those who can't see the images. That way the images are an enhancement (and not vital to understanding the concepts for people who can't refer to them).

**Comment by kapunahelewong** [comment]
About the error loading the image, could it be that the extension of the file is all caps?

File: aio/content/guide/providers.md (hunk 1 above)
Comment (santoshyadav198613) [reply aio/content/guide/providers.md]: Done.

File: aio/content/guide/providers.md (hunk 2 above)
Comment (santoshyadav198613) [reply aio/content/guide/providers.md]: Done and removed this text to avoid confusion.

**Comment by santoshyadav198613** [comment]
> About the error loading the image, could it be that the extension of the file is all caps?

Thanks it worked now, changed to svg.

File: aio/content/guide/providers.md [hunk 3]
```diff
@@ -52,6 +52,14 @@ Any component created within a lazy loaded module’s context, such as by router
 
 Though you can provide services by lazy loading modules, not all services can be lazy loaded. For instance, some modules only work in the root module, such as the Router. The Router works with the global location object in the browser.
 
+With Angular version 9, you can provide a new instance of service with each lazy loaded module. The change to `UserService` would be as follows:
+
+<code-example path="providers/src/app/user.service.2.ts"  header="src/app/user.service.ts"></code-example>
+
+Using `providedIn: 'any'` is available for provider scope, With this configuration, all eagerly loaded modules share a singleton instance, however lazy loaded modules each get their own unique instance, as in the following diagram.
```
Comment (kapunahelewong) [inline aio/content/guide/providers.md]: How about:

> With `providedIn: 'any'`, all eagerly loaded modules share a singleton instance, however lazy loaded modules each get their own unique instance, as in the following diagram.

**Review by kapunahelewong — CHANGES_REQUESTED** [review]
Just one suggestion for the copy. Also, nice work on the diagram. I've attached a screenshot with some small notes on the diagram. Thanks so much for all your hard work, @santoshyadav198613!!
santosh-img-edits

File: aio/content/guide/providers.md (hunk 3 above)
Comment (santoshyadav198613) [reply aio/content/guide/providers.md]: Made some changes to image and changes the text.

File: aio/content/guide/providers.md [hunk 4]
```diff
@@ -52,6 +52,14 @@ Any component created within a lazy loaded module’s context, such as by router
 
 Though you can provide services by lazy loading modules, not all services can be lazy loaded. For instance, some modules only work in the root module, such as the Router. The Router works with the global location object in the browser.
 
+With Angular version 9, you can provide a new instance of service with each lazy loaded module. The change to `UserService` would be as follows:
```
Comment (kapunahelewong) [inline aio/content/guide/providers.md]: Oops! I must have missed this one:
```suggestion
With Angular version 9, you can provide a new instance of a service with each lazy loaded module. The change to `UserService` would be as follows:
```

**Review by kapunahelewong — COMMENTED** [review]
It's looking great! One small comment and a question - why did you choose to move the `UserService` boxes inside the lazy loaded modules box? It looked good outside, like the one above it. If they're different, we may need to explain it/point it out (with only a sentence or two).

**Comment by santoshyadav198613** [comment]
> xplain it/point it out (with only a sentence or two)

I did it because the box is entire application, the image denotes that entire application uses same instance, and another image to show, that lazy loaded module is part of application and shares the unique instance per module.

**Comment by santoshyadav198613** [comment]
Hi @kapunahelewong ,
Can you review please.

**Comment by santoshyadav198613** [comment]
Hi @kapunahelewong ,
Can you review please.

**Comment by santoshyadav198613** [comment]
Hi Team,
Can you review it, please?

File: aio/content/guide/providers.md [hunk 5]
```diff
@@ -52,6 +52,14 @@ Any component created within a lazy loaded module’s context, such as by router
 
 Though you can provide services by lazy loading modules, not all services can be lazy loaded. For instance, some modules only work in the root module, such as the Router. The Router works with the global location object in the browser.
 
+With Angular version 9, you can provide a new instance of a service with each lazy loaded module. The change to `UserService` would be as follows:
```
Comment (aikidave) [inline aio/content/guide/providers.md:55]:
```suggestion
With Angular version 9, you can provide a new instance of a service with each lazy loaded module. The following code adds this functionality to `UserService`.
```

File: aio/content/guide/providers.md [hunk 6]
```diff
@@ -52,6 +52,14 @@ Any component created within a lazy loaded module’s context, such as by router
...
+
+<code-example path="providers/src/app/user.service.2.ts"  header="src/app/user.service.ts"></code-example>
+
+With `providedIn: 'any'`, all eagerly loaded modules share a singleton instance, however lazy loaded modules each get their own unique instance, as in the following diagram.
```
Comment (aikidave) [inline aio/content/guide/providers.md:59]:
```suggestion
With `providedIn: 'any'`, all eagerly loaded modules share a singleton instance; however, lazy loaded modules each get their own unique instance, as shown in the following diagram.
```

**Comment by santoshyadav198613** [comment]
Thanks for review @aikidave , all changes are done.

**Comment by santoshyadav198613** [comment]
Hi @aikidave , can you check now? let me know any changes are required.

**Comment by santoshyadav198613** [comment]
Hi, @alxhub @IgorMinar can you review, please.

**Review by kapunahelewong — APPROVED** [review]
Just realized I forgot to _submit_ my teeny review! Thank you, @santoshyadav198613!

Reviewed-for: global-docs-approvers

**Comment by santoshyadav198613** [comment]
Hi @kapunahelewong ,
Thanks for review done with the changes.

**Review by IgorMinar — APPROVED** [review]
LGTM. Thanks!

Reviewed-for: code-ownership

**Review by IgorMinar — COMMENTED** [review]
btw @santoshyadav198613 I recommend that for the diagram we put each of the lazy loaded chunks into each own blue box. that way it doesn't look like they are all loaded together, and the diagram will be easier to digest.

**Comment by santoshyadav198613** [comment]
> btw @santoshyadav198613 I recommend that for the diagram we put each of the lazy loaded chunks into each own blue box. that way it doesn't look like they are all loaded together, and the diagram will be easier to digest.

Sure will make the changes.

**Comment by santoshyadav198613** [comment]
@IgorMinar please check now.

**Comment by IgorMinar** [comment]
Hmm. Looks the same to me. Can you check?

**Comment by santoshyadav198613** [comment]
> Hmm. Looks the same to me. Can you check?

Oops my bad, misunderstood your reply. I will make the changes.

**Review by aikithoughts — APPROVED** [review]

**Comment by santoshyadav198613** [comment]
Hi @IgorMinar let me know if there is any changes needed

**Comment by santoshyadavdev** [comment]
Hi Team,
Let me know if any changes are needed.

**Comment by fmalcher** [comment]
Thank you @santoshyadavdev !
I still have som very minor concerns about the image:
- it's SVG but has no vector paths in it but just embedded PNG
- it's higher than wide so it could (!) look strange on the docs page in full width. IMO we could make it a bit wider so it looks a bit more asthetic
- some of the labels in the boxes are not aligned to the center

Which tool did you use to draw them? Let me know if I can help. I can redraw the image as SVG if you like. 🙂

**Comment by santoshyadavdev** [comment]
Thank @fmalcher, yes I am very bad at creating images 😅 I used https://vectr.com/ to create this one. Would be great help if you can help with correct SVG 😊

**Comment by fmalcher** [comment]
https://www.dropbox.com/s/og8e1p5qebrt47k/any-provider.svg?dl=1

What do you think? @santoshyadavdev @kapunahelewong

**Comment by santoshyadavdev** [comment]
> Screen Shot 2020-07-08 at 19 52 12
> 
> https://www.dropbox.com/s/og8e1p5qebrt47k/any-provider.svg?dl=1
> 
> What do you think? @santoshyadavdev @kapunahelewong

Much better than what I had :)

**Comment by fmalcher** [comment]
great! use this minified SVG then, please 😅 https://www.dropbox.com/s/kfn6o2wbaxii0im/any-provider.svg?dl=0

**Comment by santoshyadavdev** [comment]
> great! use this minified SVG then, please 😅 https://www.dropbox.com/s/kfn6o2wbaxii0im/any-provider.svg?dl=1
> @santoshyadavdev

Thanks @fmalcher Just a question I am getting a watermark in the image saying **Viewer does not support full svg 1.1**
image

**Comment by fmalcher** [comment]
> > great! use this minified SVG then, please 😅 https://www.dropbox.com/s/kfn6o2wbaxii0im/any-provider.svg?dl=1
> > @santoshyadavdev
> 
> Thanks @fmalcher Just a question I am getting a watermark in the image saying **Viewer does not support full svg 1.1** its not there in original image though
> image

Oops, got it... use this version then – and let me know if it doesn't work yet!
https://www.dropbox.com/s/og8e1p5qebrt47k/any-provider_.svg?dl=1

**Comment by santoshyadavdev** [comment]
> > > great! use this minified SVG then, please 😅 https://www.dropbox.com/s/kfn6o2wbaxii0im/any-provider.svg?dl=1
> > > @santoshyadavdev
> > 
> > 
> > Thanks @fmalcher Just a question I am getting a watermark in the image saying **Viewer does not support full svg 1.1** its not there in original image though
> > image
> 
> Oops, got it... use this version then – and let me know if it still doesn't work!
> https://www.dropbox.com/s/og8e1p5qebrt47k/any-provider_.svg?dl=1

Yes, this worked thanks :)

**Comment by santoshyadavdev** [comment]
Thanks a lot, @fmalcher Pushed with the new image 👍

**Review by IgorMinar — APPROVED** [review]
looks good now. thanks for updating the diagram @santoshyadavdev and thanks for helping out @fmalcher ❤️

**Pull request closed without merging**